streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java

```diff
@@ -325,6 +334,12 @@ public void checkpoint(final Map<TopicPartition, Long> ackedOffsets) {
         }
         // write the checkpoint file before closing, to indicate clean shutdown
         try {
+            if (checkpoint == null) {
+                assert eosEnabled;
+                checkpoint = new OffsetCheckpoint(new File(baseDir, CHECKPOINT_FILE_NAME));
+            } else {
+                assert !eosEnabled;
```

Comment:
@mjsax I suggested taking that discussion elsewhere because it's not related to the exactly-once work and it requires:
1. Evidence that it's worth disabling assertions in production. As far as I can tell, no-one else seems to be in favour of this at the moment.
2. Infrastructure work to enable assertions in tests.